File: gaiag/runtime/c++/runtime.cc
// Dezyne --- Dezyne command line tools
//
// Copyright © 2014, 2015, 2016, 2017 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2015, 2016, 2017, 2019 Rutger van Beusekom <rutger.van.beusekom@verum.com>
// Copyright © 2015 Paul Hoogendijk <paul.hoogendijk@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include <dzn/runtime.hh>
#include <dzn/pump.hh>

#include <algorithm>
#include <iostream>

namespace dzn
{
  std::ostream debug(nullptr);

  runtime::runtime(){}

  void trace_in(std::ostream& os, port::meta const& m, const char* e)
  {
    os << path(m.requires.meta, m.requires.port) << "." << e << " -> "
       << path(m.provides.meta, m.provides.port) << "." << e;
  }
  void trace_out(std::ostream& os, port::meta const& m, const char* e)
  {
    os << path(m.provides.meta, m.provides.port) << "." << e << " -> "
       << path(m.requires.meta, m.requires.port) << "." << e;
  }

  bool runtime::external(void* scope) {
    return (queues.find(scope) == queues.end());
  }

  bool& runtime::handling(void* scope)
  {
    return std::get<0>(queues[scope]);
  }

  void*& runtime::deferred(void* scope)
  {
    return std::get<1>(queues[scope]);
  }

  std::queue<std::function<void()> >& runtime::queue(void* scope)
  {
    return std::get<2>(queues[scope]);
  }

  bool& runtime::performs_flush(void* scope)
  {
    return std::get<3>(queues[scope]);
  }

  bool& runtime::skip_block(void* port)
  {
    return skip_port[port];
  }

  void runtime::flush(void* scope)
  {
#ifdef DEBUG_RUNTIME
    std::cout << path(reinterpret_cast<dzn::meta*>(scope)) << " flush" << std::endl;
#endif
    if(!external(scope))
    {
      std::queue<std::function<void()> >& q = queue(scope);
      while(! q.empty())
      {
        std::function<void()> event = q.front();
        q.pop();
        handle(scope, event);
      }
      if (deferred(scope)) {
        void* tgt = deferred(scope);
        deferred(scope) = NULL;
        if (!handling(tgt)) {
          runtime::flush(tgt);
        }
      }
    }
  }

  void runtime::defer(void* src, void* tgt, const std::function<void()>& event)
  {
#ifdef DEBUG_RUNTIME
    std::cout << path(reinterpret_cast<dzn::meta*>(tgt)) << " defer" << std::endl;
#endif

    if(!(src && performs_flush(src)) && !handling(tgt))
    {
      handle(tgt, event);
    }
    else
    {
      deferred(src) = tgt;
      queue(tgt).push(event);
    }
  }
}

File: gaiag/runtime/python/runtime.py
# Dezyne --- Dezyne command line tools
# Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
# Copyright © 2015 Jan Nieuwenhuizen <jan@avatar.nl>
#
# This file is part of Dezyne.
#
# Dezyne is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# Dezyne is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public
# License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
# 
# Commentary:
# 
# Code:

import sys

class V:
    def __init__ (self, v):
        self.v = v

def illegal ():
    raise RuntimeError ('illegal')

class Runtime:
    def __init__ (self, illegal=illegal):
        self.components = []
        self.illegal = illegal
        self.info = {}
    def flushes (self, c):
        self.components += [c]
        self.info[c] = self.info.get (c, Info ())
        self.info[c].flushes = True
    def external_p (self, c):
        return c not in self.components
    def flush (self, c):
        if (self.external_p (c)):
            return
        while (self.info[c].q):
            self.handle (c, self.info[c].q.pop ())
        if (self.info[c].deferred):
            t = self.info[c].deferred
            self.info[c].deferred = None;
            if (not self.info[t].handling):
                self.flush (t)
    def defer (self, i, o, f):
        if (not i or (not self.info[i].flushes and not self.info[o].handling)):
            self.handle (o, f)
        else:
            self.info[i].deferred = o
            self.info[o].q.insert (0, f)
    def valued_helper (self, c, f, m):
        handle = self.info[c].handling
        self.info[c].handling = True
        r = f ()
        if (handle and r != None):
            throw ('a valued event cannot be deferred')
        self.info[c].handling = False
        self.flush (c)
        return r
    def handle (self, c, f):
        if (not self.info[c].handling):
            self.info[c].handling = True
            f ()
            self.info[c].handling = False
            self.flush (c)
        else:
            throw ('component already handling an event')

class Port:
    def __init__ (self, name='', component=None):
        self.name = name
        self.component = component

class Component:
    def __init__ (self, loc, name='', parent=None):
        self.loc = loc
        self.rt = loc.get (Runtime)
        self.name = name
        self.parent = parent

class Info:
    def __init__ (self):
        self.handling = False
        self.deferred = None
        self.flushes = False
        self.q = []

def call_in (c, f, m):
    trace_in (m[0], m[1])
    r = c.rt.valued_helper (c, f, m)
    trace_out (m[0], 'return' if r == None else m[2][r])
    return r

def call_out (c, f, m):
    trace_out (m[0], m[1])
    c.rt.defer (m[0].inport.component, c, f)

def path (m, p=''):
    if (not m):
        return '<external>.' + p;
    if ('component' in m.__dict__.keys ()):
        return path (m.component, m.name + ('.' + p if p else p))
    if ('parent' in m.__dict__.keys () and m.parent):
        return path (m.parent, m.name + ('.' + p if p else p))
    return m.name + ('.' + p if p else p)

def trace_in (i, e):
    sys.stderr.write (path (i.outport) + '.' + e + ' -> '
                      + path (i.inport) + '.' + e + '\n')

def trace_out (i, e):
    sys.stderr.write (path(i.inport) + '.' + e + ' -> '
                      + path(i.outport) + '.' + e + '\n')
